client.html

```diff
@@ -53,6 +78,221 @@
+                    }
+                }
+            } else {
+                var newViewportYmax = elem1y;
+                var newViewportYmin = elem1y + elem1height - viewportHeight;
+                // "ideal" viewport dimensions for showing all elements in one view
+                var idealViewportY = elemMinY;
+                var idealViewportHeight = elemMaxY + elemMaxYheight - elemMinY;
+                // translate the viewport around elem1y according to location of other elements
+                newViewportY = elem1y + (idealViewportY - elem1y) / idealViewportHeight * viewportHeight;
+            }
+
+            ASMDW.scrollYtarget = newViewportY;
+            ASMDW.scrollYlastUpdate = performance.now();
+        }
+
+        function highlightBranchClearAll() {
+            var elems = document.getElementsByClassName('branch-indicator-wrapper');
+            for (var i = 0; i < elems.length; i++) {
+                var elem = elems[i];
+                elem.classList.remove('branch-highlight-stay');
+                elem.classList.remove('branch-highlight-temp');
+            }
+        }
+
+        function highlightBranchStay(className) {
+            var elems = document.getElementsByClassName(className);
+            for (var i = 0; i < elems.length; i++) {
+                var elem = elems[i];
+                elem.parentNode.classList.add('branch-highlight-stay');
+            }
+        }
+
+        function highlightBranchTemp(className) {
+            var elems = document.getElementsByClassName(className);
+            for (var i = 0; i < elems.length; i++) {
+                var elem = elems[i];
+                elem.parentNode.classList.add('branch-highlight-temp');
+            }
+        }
+
+        function highlightBranchTempClear(className) {
+            var elems = document.getElementsByClassName(className);
+            for (var i = 0; i < elems.length; i++) {
+                var elem = elems[i];
+                elem.parentNode.classList.remove('branch-highlight-temp');
+            }
+        }
+
+        function onClickBranchOrigin(ev) {
+            scrollIntoView(document.getElementById(ev.target.dataset.branchTarget), [ev.target]);
+            highlightBranchClearAll();
+            highlightBranchStay(ev.target.dataset.branchesClass);
+        }
+
+        function onClickBranchTarget(ev) {
+            // get all branch origins that branch to this branch target (ev.target)
+            var branchIndicatorsElems = document.getElementsByClassName(ev.target.dataset.branchesClass);
+            var branchOriginsElems = [];
+            for (var i = 0; i < branchIndicatorsElems.length; i++) {
+                if (branchIndicatorsElems[i] != ev.target) {
+                    branchOriginsElems.push(branchIndicatorsElems[i]);
+                }
+            }
+
+            // use a different origin as main scroll-to element each click
```

Comment:
hm, I think this behavior might be undiscoverable and confusing when you trigger it by accident. may also not be necessary with persistent highlighting?

Reply:
Agreed, see c93db8875e27e635af66af97fce11657db652666 : I replaced this by asking the user through a popup div if the branch spots end up being too far to show all of them at once

screenshots:
<details>

How it looks with enough space:
![](https://421.es/doyu/23ec05)
With not enough space, clicking on the target shows:
![](https://421.es/doyu/23ebyw)

</details>